zlib: fix raw inflate with custom dictionary

thusoy opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
zlib


##### Description of change
<!-- Provide a description of the change below this comment. -->

Moves inflateSetDictionary right after inflateInit2 when mode is
INFLATERAW, since without the wrapper in appears zlib won't return
Z_NEED_DICT as it would otherwise, and will thus attempt inflating
without the dictionary, leading to an error.

Closes #8507.

addaleax commented on src/node_zlib.cc:
```diff
@@ -517,12 +523,28 @@ class ZCtx : public AsyncWrap {
+      case INFLATERAW:
+        ctx->err_ = inflateInit2(&ctx->strm_, ctx->windowBits_);
+        ctx->env()->isolate()
+            ->AdjustAmountOfExternalAllocatedMemory(kInflateContextSize);
```
Just a suggestion, but if you want to avoid duplicating this, you can just turn the `break;` statement above into `if (ctx->mode_ != INFLATERAW) break;` and leave the `case INFLATERAW:` where it is.

addaleax replied:
Even better: Could this be moved into `SetDictionary`?

addaleax commented on src/node_zlib.cc:
```diff
@@ -517,12 +523,28 @@ class ZCtx : public AsyncWrap {
+                                           ctx->dictionary_len_);
+          if (ctx->err_ != Z_OK) {
+            // We can't detect mismatching dictionaries in raw mode since we
+            // don't have the dictid from the wrapper, thus the only failures
```
Again, just a suggestion, but using personal pronouns in the code/comments is generally discouraged around here.

addaleax commented on test/parallel/test-zlib-dictionary-raw.js:
```diff
@@ -0,0 +1,89 @@
+  });
+
+  inflate.on('data', function(chunk) {
+    output += chunk;
```
If you concatenate strings here, you should probably call `inflate.setEncoding('utf-8');` before the `on('data')` call… that might not make a difference for the single-byte-only characters here but I’d say it’s good practice to do that.

addaleax commented on test/parallel/test-zlib-dictionary-raw.js:
```diff
@@ -0,0 +1,89 @@
+  });
+
+  inflate.on('end', function() {
+    assert.equal(input, output);
```
This should probably be `assert.strictEqual()`.

addaleax commented on test/parallel/test-zlib-dictionary-fail.js:
```diff
@@ -26,3 +26,17 @@ var zlib = require('zlib');
   stream.write(Buffer.from([0x78, 0xBB, 0x04, 0x09, 0x01, 0xA5]));
 }
+
+// Should raise an error, not trigger an assertion in src/node_zlib.cc
```
This doesn’t seem to trigger an assertion right now?

addaleax commented:
This looks awesome so far, thanks for putting this together!

thusoy commented:
Thanks for the feedback! I'll address the points you raised tonight.

thusoy commented:
@addaleax I moved the dictionary initialization into SetDictionary like you suggested as it was much cleaner, thanks! Since the extra test file I added was nearly identical to the existing dictionary test file I fixed the same issues you raised in the new one in the old one as well. Do you think I should merge these two files since they are so similar, or should they be kept separate?

addaleax commented on test/parallel/test-zlib-dictionary-raw.js:
```diff
@@ -0,0 +1,91 @@
+    inflate.end();
+  });
+
+  inflate.on('end', function() {
```
Could you wrap all of these functions which will be called exactly once in a `common.mustCall()`? (You can do that in `test-zlib-dictionary.js` too, if you want)

addaleax commented:
> Since the extra test file I added was nearly identical to the existing dictionary test file I fixed the same issues you raised in the new one in the old one as well. Do you think I should merge these two files since they are so similar, or should they be kept separate?

That’s your call, if you think it’s easy to merge them together, why not?

This looks good to me (LGTM) with `common.mustCall()` and a green CI.

Here’s a CI run: https://ci.nodejs.org/job/node-test-commit/5048/

thusoy commented:
@addaleax Merged the files and added `common.mustCall` to the callbacks.

addaleax commented on test/parallel/test-zlib-dictionary.js:
```diff
@@ -58,23 +60,84 @@ function deflateResetDictionaryTest() {
+  deflate.on('data', common.mustCall(function(chunk) {
+    if (doneReset)
+      inflate.write(chunk);
+  }, 2));
```
I’d say `common.mustCall()` for `data` callbacks is optional, I don’t think we’re currently guaranteeing anything about the exact chunking or want to do that (I think this is okay if the CI is green with it included, so it’s your call).

thusoy replied:
Fair point, seems like a leaky implementation detail. I removed the `mustCall`s from non-assertion functions, as if the assertion passes the others are guaranteed to have run, and it will fail if the assertion isn't called.

thusoy commented:
Pushed new (final?) changeset without `mustCall` for non-assertion functions. Thanks for the feedback so far @addaleax.

addaleax commented:
Eek, merge conflicts… @thusoy can you rebase this? Apparently one of thea general cleanup PRs of the tests here got in the way of your own cleanup… sorry for that!

thusoy commented:
@addaleax Yup, done.

addaleax commented:
That was fast! Still LGTM.

New CI: https://ci.nodejs.org/job/node-test-commit/5189/

I’m going to land this if it’s green.

addaleax commented:
Landed in c775514e56e5e20e93c168e679b84c13f2e739b5 & 2b5acda7a297c785e3578b5dc0e4ee1743f38bd7. Thank you, this was a great contribution!

thusoy commented:
Great stuff, thanks @addaleax! Happy to contribute!

TheAlphaNerd commented:
backport?

addaleax commented:
@TheAlphaNerd If that is a yes/no question, yes. :)

@thusoy Would you be up for backporting this change to the v4.x release line? There are some conflicts in the test files that seem to need resolving.

The usual process for that would be checking out the `v4.x-staging` branch here and cherry-picking the commits as they landed on master (only c775514 should suffice), then resolving any conflicts and PRing with the staging branch as the base branch. I or somebody else can do that too, but I’m a bit busy at the moment and you clearly did a good job here in the first place.

thusoy commented:
@addaleax Thanks for the heads-up on the process. I might be able to that, but will probably be a couple of days before I have time. Consider me on the case for now.
